src/ldp.rs

```diff
@@ -1072,6 +1111,199 @@ impl ProofSuite for Eip712Signature2021 {
+            // Algorithm not supporting public key recovery: put public key in proof if it is not
+            // in the VM.
+            // TODO: dereference VM URL to check if VM contains public key.
+            let jwk_value = serde_json::to_value(public_key.to_public())?;
+            property_set.insert("publicKeyJwk".to_string(), jwk_value);
+        }
+        let proof = Proof {
+            context: TZVM_CONTEXT.clone(),
+            proof_purpose: options.proof_purpose.clone(),
+            verification_method: options.verification_method.clone(),
+            created: Some(options.created.unwrap_or_else(now_ms)),
+            domain: options.domain.clone(),
+            challenge: options.challenge.clone(),
+            property_set: Some(property_set),
+            ..Proof::new("TezosSignature2021")
+        };
+        let micheline = micheline_from_document_and_options(document, &proof).await?;
+        let micheline_string = hex::encode(micheline);
+        Ok(ProofPreparation {
+            proof,
+            jws_header: None,
+            signing_input: SigningInput::Micheline {
+                micheline: micheline_string,
+            },
+        })
+    }
+
+    async fn complete(
+        &self,
+        preparation: ProofPreparation,
+        signature: &str,
+    ) -> Result<Proof, Error> {
+        let mut proof = preparation.proof;
+        proof.proof_value = Some(signature.to_string());
+        Ok(proof)
+    }
+
+    async fn verify(
+        &self,
+        proof: &Proof,
+        document: &(dyn LinkedDataDocument + Sync),
+        resolver: &dyn DIDResolver,
+    ) -> Result<(), Error> {
+        let sig_bs58 = proof
+            .proof_value
+            .as_ref()
+            .ok_or(Error::MissingProofSignature)?;
+        let verification_method = proof
+            .verification_method
+            .as_ref()
+            .ok_or(Error::MissingVerificationMethod)?;
+        let proof_jwk_opt: Option<JWK> = match proof.property_set {
+            Some(ref props) => match props.get("publicKeyJwk") {
+                Some(jwk_value) => serde_json::from_value(jwk_value.clone())?,
+                None => None,
+            },
+            None => None,
+        };
+        let sig = bs58::decode(&sig_bs58).with_check(None).into_vec()?;
+        let (algorithm, sig) = if sig_bs58.starts_with("edsig") {
+            (Algorithm::EdDSA, sig[5..].to_vec())
+        } else if sig_bs58.starts_with(SPSIG) {
+            let alg = if proof_jwk_opt.is_some() {
+                Algorithm::ES256K
+            } else {
+                Algorithm::ES256KR
+            };
+            (alg, sig[5..].to_vec())
+        } else if sig_bs58.starts_with("p2sig") {
+            (Algorithm::ES256, sig[4..].to_vec())
+        } else {
+            return Err(Error::SignaturePrefix);
+        };
+
+        let vm = resolve_vm(&verification_method, resolver).await?;
+        if vm.type_ != "TezosMethod2021" {
```

Comment:
A VM URL doesn't necessarily contain a fragment, although in our current uses they do. But a VM could theoretically be at a DID URL using a path instead of a fragment. Relevant issue: https://github.com/w3c/did-core/issues/708

I think even a DID document itself could be a verification method, as seen in this DID document that uses an empty string (relative URI reference that resolves to the base URI, i.e. the DID document), and JSON-LD supports cyclical data structures: https://github.com/transmute-industries/vc.transmute.world/blob/master/packages/vc-http-api/src/data/didDocument.json#L84. But `resolve_vm` is not trying to support this usage.

Also related: `LinkedDataProofs::sign` currently requires a VM URL for TezosMethod2021 is required to end with `"#TezosMethod2021"`, but this is a hack which should be fixed by refactoring `ldp.rs`.

Step 4 of [derefencing the primary resource](https://w3c-ccg.github.io/did-resolution/#dereferencing-algorithm-primary) should be reached if step 3 falls through (the DID method/resolver doesn't implement `dereference` or returns None for it, and we don't have any client-/application-specific handling):
https://github.com/spruceid/ssi/blob/main/src/did_resolve.rs#L458